pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/JsonStatementOptimizer.java

```diff
@@ -0,0 +1,561 @@
+      if (expression.getType() == ExpressionType.FUNCTION && !expression.getFunctionCall().getOperator().equals("AS")
+          && result.getSecond()) {
+        // Since this is not an AS function (user-specified alias) and the function or its arguments contain json path
+        // expression, set an alias for the expression after replacing json path expression with JSON_EXTRACT_SCALAR
+        // function.
+        Function aliasFunction = getAliasFunction(result.getFirst(), expression.getFunctionCall());
+        expression.setFunctionCall(aliasFunction);
+      }
+    }
+
+    // In WHERE clause, replace JSON path expressions with JSON_MATCH function.
+    Expression filter = query.getFilterExpression();
+    if (filter != null) {
+      optimizeJsonPredicate(filter, config, schema);
+    }
+
+    // In GROUP BY clause, replace JSON path expressions with JSON_EXTRACT_SCALAR function without an alias.
+    expressions = query.getGroupByList();
+    if (expressions != null) {
+      for (Expression expression : expressions) {
+        optimizeJsonIdentifier(expression, schema, DataSchema.ColumnDataType.STRING);
+      }
+    }
+
+    // In ORDER BY clause, replace JSON path expression with JSON_EXTRACT_SCALAR. This expression must match the
+    // corresponding SELECT list expression except for the alias.
+    expressions = query.getOrderByList();
+    if (expressions != null) {
+      for (Expression expression : expressions) {
+        optimizeJsonIdentifier(expression, schema, DataSchema.ColumnDataType.STRING);
+      }
+    }
+
+    // In HAVING clause, replace JSON path expressions with JSON_EXTRACT_SCALAR. This expression must match the
+    // corresponding SELECT list expression except for the alias.
+    Expression expression = query.getHavingExpression();
+    if (expression != null) {
+      optimizeJsonIdentifier(expression, schema, DataSchema.ColumnDataType.STRING);
+    }
+  }
+
+  /**
+   * Replace an json path expression with an aliased JSON_EXTRACT_SCALAR function.
+   * @param expression input expression to rewrite into JSON_EXTRACT_SCALAR function if the expression is json path.
+   * @param outputDataType to keep track of output datatype of JSON_EXTRACT_SCALAR function which depends upon the outer
+   *                 function that json path expression appears in.
+   * @return A {@link Pair} of values where the first value is alias for the input expression and second
+   * value indicates whether json path expression was found (true) or not (false) in the expression.
+   */
+  private static Pair<String, Boolean> optimizeJsonIdentifier(Expression expression, @Nullable Schema schema,
+      DataSchema.ColumnDataType outputDataType) {
+    switch (expression.getType()) {
+      case LITERAL:
+        return new Pair<>(getLiteralSQL(expression.getLiteral(), true), false);
+      case IDENTIFIER: {
+        String[] parts = getIdentifierParts(expression.getIdentifier());
```

Comment:
Before extracting the parts from the identifier, we should first check whether the identifier is a column (note that we allow column name with `.` as of now)

Reply:
Looking into this.